edx_toggles/tests/test_state.py

```diff
@@ -0,0 +1,133 @@
+        )
+
+    def test_response_with_new_setting_dict_toggle(self):
+        _toggle = SettingDictToggle(
+            "CUSTOM_FEATURES", "MYSETTING", default=False, module_name="module1"
+        )
+        with override_settings(CUSTOM_FEATURES={"MYSETTING": True}):
+            report = ToggleStateReport().as_dict()
+
+        setting_dict = {toggle["name"]: toggle for toggle in report["django_settings"]}
+
+        self.assertEqual(
+            {
+                "name": "CUSTOM_FEATURES['MYSETTING']",
+                "is_active": True,
+                "module": "module1",
+                "class": "SettingDictToggle",
+            },
+            setting_dict["CUSTOM_FEATURES['MYSETTING']"],
+        )
+
+    def test_setting_overridden_by_setting_toggle(self):
+        _toggle2 = SettingToggle("MYSETTING2", module_name="module1")
+        _toggle3 = SettingDictToggle("MYDICT", "MYSETTING3", module_name="module1")
+        with override_settings(
+            MYSETTING1=True, MYSETTING2=False, MYDICT={"MYSETTING3": False}
+        ):
+            # Need to pre-load settings, otherwise they are not picked up by the view
+            self.assertTrue(settings.MYSETTING1)
+            report = ToggleStateReport().as_dict()
+
+        setting_dict = {toggle["name"]: toggle for toggle in report["django_settings"]}
+
+        # Check that Django settings for which a SettingToggle exists have both the correct is_active and class values
+        self.assertTrue(setting_dict["MYSETTING1"]["is_active"])
+        self.assertNotIn("class", setting_dict["MYSETTING1"])
+        self.assertFalse(setting_dict["MYSETTING2"]["is_active"])
+        self.assertEqual("SettingToggle", setting_dict["MYSETTING2"]["class"])
+        self.assertFalse(setting_dict["MYDICT['MYSETTING3']"]["is_active"])
+        self.assertEqual(
+            "SettingDictToggle", setting_dict["MYDICT['MYSETTING3']"]["class"]
+        )
+
+    def test_no_duplicate_setting_toggle(self):
+        _toggle1 = SettingToggle("MYSETTING1", module_name="module1")
+        _toggle2 = SettingDictToggle("MYDICT", "MYSETTING2", module_name="module1")
+        with override_settings(MYSETTING1=True, MYDICT={"MYSETTING2": False}):
+            report = ToggleStateReport().as_dict()
+
+        # Check there are no duplicate setting/toggle
+        response_toggles_1 = [
+            toggle
+            for toggle in report["django_settings"]
+            if toggle["name"] == "MYSETTING1"
+        ]
+        response_toggles_2 = [
+            toggle
+            for toggle in report["django_settings"]
+            if toggle["name"] == "MYDICT['MYSETTING2']"
+        ]
+        self.assertEqual(1, len(response_toggles_1))
+        self.assertEqual(1, len(response_toggles_2))
+
+    def test_code_owners_without_module_information(self):
+        # Create a waffle flag without any associated module_name
+        waffle_flag = WaffleFlag("test.flag2", module_name=None)
```

Comment:
Not part of this PR, but I thought module name with None was no longer allowed? I guess we don’t have an assertion?

Reply:
The `module_name` argument has to be defined, but there is no value checking. My personal opinion is that we should not be adding assert statements.

Reply:
I'll explain why I would add this, and then you can explain why you wouldn't.  I want `__name__` to be used everywhere, and allowing `None` enables people to not do this, and leaves the potential of a contagion of copy/paste with a bad example. The assertion could clearly note how to resolve. I know you didn't want this to be a required field to begin with, and I feel like this may just be a continuation of the same original discussion. I also may be missing something, so let me know if that is the case.  Thank you!

Reply:
Python allows developers to pass arbitrary values as function arguments. Ensuring argument correctness is difficult, and I believe it should be the developers' responsibility to make sure that the arguments that they use are correct. In this particular case, how hard should we try to ensure correctness?

- `module_name` is "truish" (`if not module_name: ...`)
- `module_name` is a string (`if not isinstance(module_name, str): ...`)
- `module_name` is an importable module (`try importlib.import_module(module_name): ... except ModuleNotFoundError: ...`)

Let's face it: there is no way to prevent developers from bad copy-pasting. If we write `WaffleFlag(..., "testmodule1")` in a test case, then other users might copy-paste it.

Reply:
I was thinking truish with a message of what we expect someone to use. I understand we can’t prevent every issue, but I would have prevented this. I can check what other think. I might be too much of a control freak. 🙂

I hope it is clear that this is not a blocker to this PR which has been approved. Are you waiting on me for anything for merging?

Reply:
> I hope it is clear that this is not a blocker to this PR which has been approved. Are you waiting on me for anything for merging?

You made it clear but I did not realize it :) I'll merge right away. I won't add the module_name truish check to this PR, as this would be a backward-incompatible change.

Reply:
Note: this line could still be updated to use `__name__` for all the reasons I wanted to make this required. I'm not going to block on this of course, but if you are willing, you can update. Either way, you can respond and mark resolved again.  Thanks.

Reply:
Right! I updated the argument.